Fix allocation and health checks in IPv6 only mode

tgraf opened the pull request:
Several hard dependencies on IPv4 have found their way into the codebase. Allow running in IPv6 mode again without allocating any IPv4 addresses.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6710)
<!-- Reviewable:end -->

aanm commented on daemon/daemon.go:
```diff
@@ -1049,6 +1042,13 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
 	log.Info("Initializing IPAM")
 	ipam.Init()
 
+	wOpts := map[string]string{
```
Add a comment stating this can not be moved before the IPAM initialization

aanm commented on pkg/ipam/allocator.go:
```diff
@@ -79,34 +80,54 @@ func (ipam *IPAM) AllocateIPString(ipAddr string) error {
-		ipv6 = ipConf
-		metrics.IpamEvent.WithLabelValues(metricAllocate, familyIPv6).Inc()
 	}
 
```
in this line it needs a :
```
defer func(){
 if err != nil {
    ipam.ReleaseIPv6Address(ipv6)
  }
}
```

tgraf replied:
This is an existing bug. I'll fix it in a separate so we can backport.

tgraf replied:
https://github.com/cilium/cilium/pull/6719

aanm commented on pkg/ipam/ipam.go:
```diff
@@ -49,52 +47,15 @@ const (
-			},
-		},
-		&cniTypes.Route{
-			Dst: defaults.IPv4DefaultRoute,
```
Where is the gateway defined now? Does this means that `github.com/containernetworking/plugins/plugins/ipam/host-local/backend/allocator` is not used anywhere?

tgraf replied:
I didn't find any code using this. I will double check again.

tgraf replied:
Removed the vendor code

joestringer commented on pkg/ipam/ipam.go:
```diff
@@ -0,0 +1,196 @@
+
+const (
+	IPv6 Family = "ipv6"
+	IPv4 Family = "ipv4"
```
These appear to duplicate `familyIPv6` and `familyIPv4`.

tgraf replied:
Yes, I did not want to export untyped constant and I didn't want to refactor all code either and keep the changes small.

joestringer commented on daemon/daemon.go:
```diff
@@ -1123,16 +1123,25 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
+		if option.Config.EnableIPv6 {
+			health6, err := d.ipam.AllocateNextFamily(ipam.IPv6)
+			if err != nil {
+				return nil, restoredEndpoints, fmt.Errorf("unable to allocate health IPs: %s,see https://cilium.link/ipam-range-full", err)
```
There's no de-allocation of the potentially allocated `health4` address in this error path. (Though admittedly it's probably less likely you'll run out of IPv6 addresses than IPv4 addresses and actually hit this..)

tgraf replied:
Yes, this is true for all allocations and a lot of other resources. It is currently assumed that if `NewDaemon()` fails, then the agent process is quit. I can fix it up but it doesn't really solve the overall problem.

joestringer reviewed (changes requested):
I posted few minor comments below, one unlikely bug and one duplicate set of symbols.

I also noticed that the node element model expects IPv4 AND IPv6 addresses, which is a softer dependency not handled in this PR:
https://github.com/cilium/cilium/blob/17f945350af127ed58aa105487140509fad7d646/pkg/node/node.go#L199-L221

tgraf committed:
agent: Fail if loopback IPv4 address cannot be allocated

Make sure that allocation fails if the allocator is not available

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
health: Only allocate health IP if address family is enabled

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Remove unused ipv4HealthAddress and ipv6HealthAddress

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Add AllocateNextFamily() to allocate IPs of individual families

AllocateNext() allows to allocate for a specific family but the allocation does
not fail if the allocator is unavailable.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Stop using global ipamConfig variable

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
vendor: Remove now unused dependency code

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
workloads: Fix init race condition between workloads and IPAM init

The workloads system requires IP allocation. Initialize the workloads system
after initializing IPAM.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Remove unused IPAMConfig

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
ipam: Only allocate IPs if address family is enabled

Only initialize the IP allocator for a particular address family if that
address family is actually enabled.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
health: Support running IPv6 only health endpoints

The cleanup was limited to IPv4 only so far

Signed-off-by: Thomas Graf <thomas@cilium.io>
